In [13]:
# Import required libraries for data loading and inspection
import pandas as pd
import os


In [14]:
# Set project root as a fixed absolute path
BASE = "/Users/alexia/Documents/CASA/Dissertation"
print("Project root:", BASE)
print("Exists:", os.path.exists(BASE))


Project root: /Users/alexia/Documents/CASA/Dissertation
Exists: True


## Load TS045

In [15]:
# Load Census 2021 TS045 (car/van availability by household). Nomis export has 7 metadata
# rows before the header, and a disclosure-control footer with non-numeric values in the
# first count column — both handled explicitly below (TS007A's header was clean; this one
# is not). Replaces TS007A entirely: driving-age population (Pi) is no longer used anywhere
# in the demand formula, which now runs on household vehicle stock (Hi x Ci) instead.
ts045_path = os.path.join(BASE, "03_data/demand/census2021/TS045_car_van_availability_London_LSOA_2021.csv")
ts045_raw = pd.read_csv(ts045_path, skiprows=7)

ts045_raw = ts045_raw.rename(columns={
    "Area": "area_raw",
    "No cars or vans in household": "cars_0",
    "1 car or van in household": "cars_1",
    "2 cars or vans in household": "cars_2",
    "3 or more cars or vans in household": "cars_3plus",
})

# Drop the disclosure-control footer (rows with no area_raw), and coerce count columns to
# numeric -- the footer text in cars_0 forces that column to string dtype otherwise
ts045_raw = ts045_raw.dropna(subset=["area_raw"])
for col in ["cars_0", "cars_1", "cars_2", "cars_3plus"]:
    ts045_raw[col] = pd.to_numeric(ts045_raw[col], errors="coerce")

# Keep the gor:London aggregate row separately -- used in 02_data_cleaning as an independent
# check that the London LSOA subset sums to the correct household total
ts045_gor_london = ts045_raw[ts045_raw["area_raw"] == "gor:London"].copy()

# Parse "lsoa2021:E01000001 : City of London 001A" into lsoa_code + lsoa_name.
# Despite the filename, this file is all England+Wales LSOAs, not pre-filtered to London --
# same as TS007A, the London filter happens in 02_data_cleaning, not here.
ts045_lsoa = ts045_raw[ts045_raw["area_raw"].str.startswith("lsoa2021:", na=False)].copy()
parsed = ts045_lsoa["area_raw"].str.extract(r"lsoa2021:(?P<lsoa_code>\S+)\s*:\s*(?P<lsoa_name>.+)")
ts045_lsoa["lsoa_code"] = parsed["lsoa_code"]
ts045_lsoa["lsoa_name"] = parsed["lsoa_name"]

ts045 = ts045_lsoa[["lsoa_code", "lsoa_name", "cars_0", "cars_1", "cars_2", "cars_3plus"]].reset_index(drop=True)

print("=== TS045 Car/Van Availability (raw, all England+Wales) ===")
print("Shape:", ts045.shape)
print("Columns:", ts045.columns.tolist())
ts045.head()


=== TS045 Car/Van Availability (raw, all England+Wales) ===
Shape: (35672, 6)
Columns: ['lsoa_code', 'lsoa_name', 'cars_0', 'cars_1', 'cars_2', 'cars_3plus']


,lsoa_code,lsoa_name,cars_0,cars_1,cars_2,cars_3plus
0,E01011954,Hartlepool 001A,305,388.0,208.0,62.0
1,E01011969,Hartlepool 001B,87,296.0,168.0,53.0
2,E01011970,Hartlepool 001C,48,231.0,153.0,53.0
3,E01011971,Hartlepool 001D,29,201.0,204.0,86.0
4,E01033465,Hartlepool 001F,56,282.0,323.0,79.0


## Load evse_location (replaces OpenStreetEV_GLA)

`OpenStreetEV_GLA.csv` is retired — supervisor-provided `evse_location.csv` covers the full
Location > Device > EVSE > Connector hierarchy in one already-joined file (hence the `_x`/`_y`
suffixed columns), so it fully replaces the old registry's role. Confirmed via a standalone
exploration pass before writing this cell:

- `id_x` (Location id) matches `location_id_x` in every row, and is the correct key for
  LSOA-level eⱼ: 23,013 unique Locations across 51,386 total rows (Location < Device < EVSE <
  Connector counts increase monotonically as expected).
- The usable coordinate pair is `coordinates_latitude_y` + `coordinates_longitude`. The
  Location-level `coordinates_latitude_x` / `coordinates_latitude_x.1` columns are a duplicate-
  latitude artefact in the source file (identical values in both, no true longitude pairing) —
  not used.
- `location_class` values are `On-street` / `Destination` / `En-route` / `Other` — note the
  lowercase "street", unlike the old `On-Street`. Any downstream filter must compare
  case-insensitively.
- `Source` / `operator_name` are kept this time (the old pipeline dropped them as "not used in
  current methodology") because they now feed directly into the L10 data-completeness
  limitation: zero-session EVSEs skew towards `Source == "Custom"` and towards missing
  `operator_name`.

In [16]:
# Load evse_location.csv -- restricted GLA/Zapmap EVSE registry (Sept 2025 snapshot)
# low_memory=False avoids a DtypeWarning on columns we aren't keeping anyway
# (external_evse_uid, status_schedule, directions, parking_restrictions, deleted_at_x)
evse_location_raw = pd.read_csv(os.path.join(BASE, "03_data/restricted/evse_location.csv"), low_memory=False)

# Keep only the columns used in the current methodology; drop near-100%-empty admin/media
# fields (terms_and_conditions, deleted_at_x, floor_level, status_schedule, parking_restrictions,
# directions, images, tariff_ids, suboperator_name, owner_name, format, standard, power_type,
# max_voltage, max_amperage, connector-level ids, timestamps)
cols_keep = [
    "id_x", "address", "city", "state", "location_class",
    "coordinates_latitude_y", "coordinates_longitude",
    "zapmap_device_uid", "power_band", "device_max_power",
    "evse_id", "evse_uid", "Source", "operator_name",
]
evse_location = evse_location_raw[cols_keep].copy()

# Rename to match the old OpenStreetEV_GLA naming convention wherever a direct equivalent
# exists, so downstream 02-05 code needs the smallest possible diff
evse_location = evse_location.rename(columns={
    "id_x":                    "location_id",
    "state":                   "borough",
    "location_class":          "location_category",
    "coordinates_latitude_y":  "latitude",
    "coordinates_longitude":   "longitude",
    "Source":                  "source",
})

print("=== evse_location.csv ===")
print("Shape:", evse_location.shape)
print("Columns:", evse_location.columns.tolist())
evse_location.head()


=== evse_location.csv ===
Shape: (51386, 14)
Columns: ['location_id', 'address', 'city', 'borough', 'location_category', 'latitude', 'longitude', 'zapmap_device_uid', 'power_band', 'device_max_power', 'evse_id', 'evse_uid', 'source', 'operator_name']


,location_id,address,city,borough,location_category,latitude,longitude,zapmap_device_uid,power_band,device_max_power,evse_id,evse_uid,source,operator_name
0,3SJEP0S,Dagenham,Dagenham,Barking and Dagenham,On-street,51.546391,0.165086,HRTQGU2,2. Fast (AC),43000,5c3059ba26954929b7c1020354f8c561,5c3059ba26954929b7c1020354f8c561,Custom,NaN
1,3SJEP0S,Dagenham,Dagenham,Barking and Dagenham,On-street,51.546391,0.165086,HRTQGU2,2. Fast (AC),43000,c1d31bbf504c4f63a7e9280d65cd749a,c1d31bbf504c4f63a7e9280d65cd749a,Custom,NaN
2,3SJEP0S,Dagenham,Dagenham,Barking and Dagenham,On-street,51.546391,0.165086,HRTQGU2,2. Fast (AC),43000,84813e04bb2e4c76a2e39e7e4e4bcfca,84813e04bb2e4c76a2e39e7e4e4bcfca,Custom,NaN
3,3SJEP0S,Dagenham,Dagenham,Barking and Dagenham,On-street,51.546391,0.165086,HRTQGU2,3. Rapid,50000,5c3059ba26954929b7c1020354f8c561,5c3059ba26954929b7c1020354f8c561,Custom,NaN
4,3SJEP0S,Dagenham,Dagenham,Barking and Dagenham,On-street,51.546391,0.165086,HRTQGU2,3. Rapid,50000,c1d31bbf504c4f63a7e9280d65cd749a,c1d31bbf504c4f63a7e9280d65cd749a,Custom,NaN


In [17]:
# Integrity checks carried over from the exploration pass -- confirm the two id fields
# that look redundant actually are, before relying on either one downstream
id_check = (evse_location_raw["id_x"] == evse_location_raw["location_id_x"]).all()
evse_check = (evse_location_raw["evse_id"] == evse_location_raw["evse_uid"]).all()
print(f"id_x == location_id_x for all rows: {id_check}")
print(f"evse_id == evse_uid for all rows: {evse_check}")

print("\n=== location_category value counts (raw casing -- compare case-insensitively downstream) ===")
print(evse_location["location_category"].value_counts(dropna=False))


id_x == location_id_x for all rows: True
evse_id == evse_uid for all rows: True

=== location_category value counts (raw casing -- compare case-insensitively downstream) ===
location_category
On-street      36667
Destination    11442
En-route        1642
Other           1635
Name: count, dtype: int64


## De-duplicate `evse_location` to Location level -> `osev`

In [18]:
# evse_location is at Connector grain (51,386 rows across 23,013 Locations) -- unlike the
# old OpenStreetEV_GLA export, which already was one row per Location. Reduce to one row per
# location_id, keeping the first occurrence's location-level attributes, to reproduce the old
# `osev` registry table used for has_evse_i / eJ.
osev = evse_location.drop_duplicates(subset=["location_id"]).reset_index(drop=True)

n_onstreet = (osev["location_category"].str.lower() == "on-street").sum()
print("=== osev (deduplicated to Location level) ===")
print("Shape:", osev.shape)
print(f"On-street locations: {n_onstreet:,}")
print("(cross-check: the corrected on-street eJ total under the old pipeline was 21,366 --")
print(f" new count is {n_onstreet:,}, consistent within rounding)")
osev.head()


=== osev (deduplicated to Location level) ===
Shape: (23013, 14)
On-street locations: 21,367
(cross-check: the corrected on-street eJ total under the old pipeline was 21,366 --
 new count is 21,367, consistent within rounding)


,location_id,address,city,borough,location_category,latitude,longitude,zapmap_device_uid,power_band,device_max_power,evse_id,evse_uid,source,operator_name
0,3SJEP0S,Dagenham,Dagenham,Barking and Dagenham,On-street,51.546391,0.165086,HRTQGU2,2. Fast (AC),43000,5c3059ba26954929b7c1020354f8c561,5c3059ba26954929b7c1020354f8c561,Custom,NaN
1,DGGPFSM,Dagenham,Dagenham,Barking and Dagenham,On-street,51.539635,0.127792,S5VVS2W,2. Fast (AC),43000,cf8c2d5aaa314d738e4de92976df0444,cf8c2d5aaa314d738e4de92976df0444,Custom,NaN
2,5ZGAF24,807-829 Longbridge Road,Dagenham,Barking and Dagenham,Destination,51.552335,0.113204,WAVDMKX,2. Fast (AC),22000,b4d67e86d86d04d0f942f6caeb49e7e7,b4d67e86d86d04d0f942f6caeb49e7e7,Custom,NaN
3,9E81MQO,79 Whalebone Lane,Dagenham,Barking and Dagenham,Destination,51.563726,0.146107,PNZ7HFT,2. Fast (AC),22000,381d092df68cea045d73a557c0721c3d,381d092df68cea045d73a557c0721c3d,Custom,NaN
4,21CK7GS,258-262 Heathway,Dagenham,Barking and Dagenham,Destination,51.542202,0.148508,DPN7O3K,2. Fast (AC),22000,0fce5367e3d94551d92113f8db3dbaad,0fce5367e3d94551d92113f8db3dbaad,Custom,NaN


## IoD2025.csv peek

In [19]:
# Peek at IoD2025 File 7 -- only need to confirm exact column names before selecting/renaming
with open(os.path.join(BASE, "03_data/demand/IoD2025.csv"), encoding="utf-8") as f:
    for i, line in enumerate(f):
        print(f"Line {i}: {line.rstrip()}")
        if i >= 5:
            break

iod_raw = pd.read_csv(os.path.join(BASE, "03_data/demand/IoD2025.csv"), nrows=3)
print("\nShape:", iod_raw.shape)
print("Columns:", iod_raw.columns.tolist())


Line 0: LSOA code (2021),LSOA name (2021),Local Authority District code (2024),Local Authority District name (2024),Index of Multiple Deprivation (IMD) Score,Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived),Index of Multiple Deprivation (IMD) Decile (where 1 is most deprived 10% of LSOAs),Income Score (rate),Income Rank (where 1 is most deprived),Income Decile (where 1 is most deprived 10% of LSOAs),Employment Score (rate),Employment Rank (where 1 is most deprived),Employment Decile (where 1 is most deprived 10% of LSOAs),"Education, Skills and Training Score","Education, Skills and Training Rank (where 1 is most deprived)","Education, Skills and Training Decile (where 1 is most deprived 10% of LSOAs)",Health Deprivation and Disability Score,Health Deprivation and Disability Rank (where 1 is most deprived),Health Deprivation and Disability Decile (where 1 is most deprived 10% of LSOAs),Crime Score,Crime Rank (where 1 is most deprived),Crime Decile (where 1 is most de

## Load IoD2025

In [20]:
# Load IoD2025 File 7 (income domain, all LSOAs in England)
iod_raw = pd.read_csv(os.path.join(BASE, "03_data/demand/IoD2025.csv"))

# Keep LSOA identifiers, composite IMD (kept for reference/comparison -- proposal discusses
# switching from composite rank to income score, useful for that narrative), and Income domain
# fields (the actual equity weight used in the demand formula)
cols_keep = [
    "LSOA code (2021)", "LSOA name (2021)",
    "Local Authority District code (2024)", "Local Authority District name (2024)",
    "Index of Multiple Deprivation (IMD) Score",
    "Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived)",
    "Index of Multiple Deprivation (IMD) Decile (where 1 is most deprived 10% of LSOAs)",
    "Income Score (rate)",
    "Income Rank (where 1 is most deprived)",
    "Income Decile (where 1 is most deprived 10% of LSOAs)",
]
iod2025 = iod_raw[cols_keep].copy()

iod2025 = iod2025.rename(columns={
    "LSOA code (2021)":                                                                      "lsoa_code",
    "LSOA name (2021)":                                                                      "lsoa_name",
    "Local Authority District code (2024)":                                                  "lad_code",
    "Local Authority District name (2024)":                                                  "lad_name",
    "Index of Multiple Deprivation (IMD) Score":                                             "imd_score",
    "Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived)":                   "imd_rank",
    "Index of Multiple Deprivation (IMD) Decile (where 1 is most deprived 10% of LSOAs)":    "imd_decile",
    "Income Score (rate)":                                                                   "income_score",
    "Income Rank (where 1 is most deprived)":                                                "income_rank",
    "Income Decile (where 1 is most deprived 10% of LSOAs)":                                 "income_decile",
})

print("=== IoD2025 File 7 ===")
print("Shape:", iod2025.shape)
print("Columns:", iod2025.columns.tolist())
iod2025.head()


=== IoD2025 File 7 ===
Shape: (33755, 10)
Columns: ['lsoa_code', 'lsoa_name', 'lad_code', 'lad_name', 'imd_score', 'imd_rank', 'imd_decile', 'income_score', 'income_rank', 'income_decile']


,lsoa_code,lsoa_name,lad_code,lad_name,imd_score,imd_rank,imd_decile,income_score,income_rank,income_decile
0,E01000001,City of London 001A,E09000001,City of London,8.742,26525,8,0.013,33730,10
1,E01000002,City of London 001B,E09000001,City of London,4.722,31203,10,0.018,33669,10
2,E01000003,City of London 001C,E09000001,City of London,9.250,25913,8,0.107,25167,8
3,E01000005,City of London 001E,E09000001,City of London,19.884,14807,5,0.211,14836,5
4,E01000006,Barking and Dagenham 016A,E09000002,Barking and Dagenham,25.307,10917,4,0.343,7519,3


## Load charging_activity_sep (replaces join_august2025)

`join_august2025.csv` is retired — supervisor-provided `charging_activity_sep.csv` covers the
full month of September 2025 (vs. the old 7-day window), a substantially better observation
window for the utilisation coefficient. Unlike the old file, this one carries **no location
attributes** — only `evse_uid`, `start_time`, `end_time`, `duration` — so it must be explicitly
joined to `evse_location` on `evse_uid` to recover location/LSOA context. That join, plus the
interval-merge and window-clip cleaning (427 EVSEs / 775 session-pairs have internal overlaps,
the same class of issue as the old ur_j > 1 bug), happens in 02_data_cleaning — this notebook
only loads and lightly renames.

In [21]:
# Peek at charging_activity_sep -- restricted GLA/Zapmap session data (Sept 2025, full month)
with open(os.path.join(BASE, "03_data/restricted/charging_activity_sep.csv"), encoding="utf-8") as f:
    for i, line in enumerate(f):
        print(f"Line {i}: {line.rstrip()}")
        if i >= 5:
            break


Line 0: evse_uid,start_time,end_time,duration
Line 1: 0002729a6be20a87309637692e5a3af3,2025/9/7 22:21,2025/9/7 22:21,0.016666666666666666
Line 2: 0002729a6be20a87309637692e5a3af3,2025/9/7 22:21,2025/9/7 23:21,59.983333333333334
Line 3: 0003081cf555fade0d74f080d8baa089,2025/9/11 21:49,2025/9/11 21:58,8.966666666666667
Line 4: 0003081cf555fade0d74f080d8baa089,2025/9/23 09:08,2025/9/23 14:29,320.8
Line 5: 00060c5bdc07754065222530a5987345,2025/9/4 17:11,2025/9/4 17:16,4.1


In [22]:
# Load charging_activity_sep.csv and parse timestamps
charging_activity_raw = pd.read_csv(os.path.join(BASE, "03_data/restricted/charging_activity_sep.csv"))

# Rename to the old join_august2025 naming convention (charging_start / charging_duration)
# wherever a direct equivalent exists, so 02-05 need the smallest possible diff.
# end_time has no old equivalent -- it's a genuine addition (old data only had start+duration).
charging_activity = charging_activity_raw.rename(columns={
    "start_time": "charging_start",
    "end_time":   "charging_end",
    "duration":   "charging_duration",
})
charging_activity["charging_start"] = pd.to_datetime(charging_activity["charging_start"])
charging_activity["charging_end"]   = pd.to_datetime(charging_activity["charging_end"])

print("=== Zapmap Sessions (charging_activity_sep) ===")
print("Shape:", charging_activity.shape)
print("Columns:", charging_activity.columns.tolist())
print("Date range:", charging_activity["charging_start"].min(), "to", charging_activity["charging_start"].max())
charging_activity.head()


=== Zapmap Sessions (charging_activity_sep) ===
Shape: (524454, 4)
Columns: ['evse_uid', 'charging_start', 'charging_end', 'charging_duration']
Date range: 2025-09-01 00:00:00 to 2025-09-30 23:59:00


,evse_uid,charging_start,charging_end,charging_duration
0,0002729a6be20a87309637692e5a3af3,2025-09-07 22:21:00,2025-09-07 22:21:00,0.016667
1,0002729a6be20a87309637692e5a3af3,2025-09-07 22:21:00,2025-09-07 23:21:00,59.983333
2,0003081cf555fade0d74f080d8baa089,2025-09-11 21:49:00,2025-09-11 21:58:00,8.966667
3,0003081cf555fade0d74f080d8baa089,2025-09-23 09:08:00,2025-09-23 14:29:00,320.800000
4,00060c5bdc07754065222530a5987345,2025-09-04 17:11:00,2025-09-04 17:16:00,4.100000


In [23]:
# Join coverage check -- carried over from the exploration pass so this notebook stands on
# its own without needing to re-run 00_new_data_exploration.ipynb
uid_in_activity = set(charging_activity["evse_uid"].unique())
uid_in_location = set(evse_location["evse_uid"].unique())
print(f"Unique evse_uid in charging_activity: {len(uid_in_activity):,}")
print(f"Unique evse_uid in evse_location:     {len(uid_in_location):,}")
print(f"Joinable (in both):                   {len(uid_in_activity & uid_in_location):,}")
print(f"In activity but not in location (should be 0): {len(uid_in_activity - uid_in_location):,}")


Unique evse_uid in charging_activity: 18,748
Unique evse_uid in evse_location:     38,358
Joinable (in both):                   18,748
In activity but not in location (should be 0): 0


## Summary check

In [24]:
# Print a summary table to confirm all datasets loaded successfully
datasets = {
    "TS045":              ts045,
    "IoD2025":            iod2025,
    "evse_location":      evse_location,
    "osev (location-lvl)": osev,
    "charging_activity":  charging_activity,
}

print(f"{'Dataset':<20} {'Rows':>12} {'Cols':>6}")
print("-" * 41)
for name, df in datasets.items():
    print(f"{name:<20} {df.shape[0]:>12,} {df.shape[1]:>6}")


Dataset                      Rows   Cols
-----------------------------------------
TS045                      35,672      6
IoD2025                    33,755     10
evse_location              51,386     14
osev (location-lvl)        23,013     14
charging_activity         524,454      4


TS045 — 筛选出London LSOA（4,994个，全量匹配gor:London汇总行），加总4个拥车档位得到Hᵢ（户数），加权平均得到Cᵢ（每户平均拥车数，3+档近似取3）

IoD2025 File 7 — 筛选London LSOA，保留income score/rank/decile（IMD composite作为新旧对照参考；需求公式里用income_score连续值，decile只用于M3评估分组）

evse_location（原OpenStreetEV_GLA + join_august2025的登记表部分）— 51,386行 / 23,013个unique location，51%的EVSE（19,610个）这个月零session，其中Custom来源零session率75.8% vs OCPI来源45.6%（新版L10限制的具体数字）。去重到location级得到`osev`（21,367个on-street location，和旧口径21,366基本一致）。用于：(1) 判断has_evse_i，(2) 提供完整EVSE总体用于Sᵢᵉᶠᶠ/Uᵢ聚合，(3) 计算eⱼ（每个LSOA现有EVSE数量，直接进p-median的容量约束C1）。

charging_activity（原join_august2025的session部分）— 九月整月524,454条session，18,748个unique EVSE，无缺失location信息（因为这次文件本身不带location属性了，必须靠evse_uid和evse_location显式join，这一步和去重+区间合并+窗口裁剪一起放到02_data_cleaning里做，01这里只加载不清洗）。窗口从7天×1,440分钟（10,080）变成整月30天×1,440分钟（43,200），ur_j分母要跟着改。